packages/llvm.rb

```diff
@@ -285,51 +53,44 @@ def self.build
-               "-DCURSES_INCLUDE_PATH='#{CREW_PREFIX}/include/ncursesw'",
-               "-DCMAKE_INSTALL_PREFIX=#{CREW_PREFIX}",
-               '-DCMAKE_CXX_FLAGS="-fPIC"',
-               '-DCMAKE_BUILD_TYPE=Release',
-               '-DLLVM_LIBDIR_SUFFIX=64',
-               '-DBUILD_SHARED_LIBS=ON',
-               '-DLLVM_ENABLE_RTTI=ON',
-               '-Wno-dev',
-               '..'
-               # Fix for file INSTALL cannot find ".../lib64/python2.7" error.
-               # See http://lists.llvm.org/pipermail/lldb-dev/2015-June/007633.html.
-               system "sed -i '40,43d' tools/lldb/cmake_install.cmake"
-      else # armv7l, aarch64 or i686
-        system 'cmake',
-               "-DCURSES_INCLUDE_PATH='#{CREW_PREFIX}/include/ncursesw'",
-               "-DCMAKE_INSTALL_PREFIX=#{CREW_PREFIX}",
-               '-DCMAKE_CXX_FLAGS="-fPIC"',
-               '-DCMAKE_BUILD_TYPE=Release',
-               '-DBUILD_SHARED_LIBS=ON',
-               '-DLLVM_ENABLE_RTTI=ON',
-               '-Wno-dev',
-               '..'
-      end
-      system 'make'
+
+        system "cmake -G Ninja \
+            -DCMAKE_INSTALL_PREFIX=#{CREW_PREFIX} \
+            -DLLVM_DEFAULT_TARGET_TRIPLE=#{ARCH}-cros-linux-gnu \
+            -DLLVM_TARGETS_TO_BUILD=#{LLVM_TARGETS_TO_BUILD} \
+            -DCMAKE_BUILD_TYPE=Release \
+            -DLLVM_LIBDIR_SUFFIX='#{CREW_LIB_SUFFIX}' \
+            -DLLVM_BUILD_LLVM_DYLIB=ON \
+            -DLLVM_ENABLE_RTTI=ON \
+            -DLLVM_USE_LINKER=gold \
+            -DLLVM_INSTALL_UTILS=ON \
+            -DLLVM_BINUTILS_INCDIR='../binutils/include' \
+            -DLLVM_ENABLE_PROJECTS='clang;clang-tools-extra;libcxx;libcxxabi;libunwind;lldb;compiler-rt;lld;polly;openmp' \
+            -Wno-dev \
+            ../llvm"
+      system 'ninja'
     end
-    system "ld_default #{original_default}"
   end
 
   def self.install
     Dir.chdir("builddir") do
       system "install -Dm755 clc #{CREW_DEST_PREFIX}/bin/clc"
       system "install -Dm755 clc++ #{CREW_DEST_PREFIX}/bin/clc++"
-      system 'make', "DESTDIR=#{CREW_DEST_DIR}", 'install'
+      system "DESTDIR=#{CREW_DEST_DIR} ninja install"
+      FileUtils.mkdir_p "#{CREW_DEST_LIB_PREFIX}/bfd-plugins"
+      #FileUtils.cp lib#{CREW_LIB_SUFFIX}/LLVMgold.so "#{CREW_DEST_LIB_PREFIX}/bfd-plugins/"
+      #FileUtils.cp lib#{CREW_LIB_SUFFIX}/libLTO.* "#{CREW_DEST_LIB_PREFIX}/bfd-plugins/"
+      system "cp lib#{CREW_LIB_SUFFIX}/LLVMgold.so #{CREW_DEST_LIB_PREFIX}/bfd-plugins/"
+      system "cp lib#{CREW_LIB_SUFFIX}/libLTO.* #{CREW_DEST_LIB_PREFIX}/bfd-plugins/"
     end
   end
   
   #def self.check
-  #  Dir.chdir("builddir") do
-  #    system "make -k -j#{CREW_NPROC} check-llvm"
-  #    system "make -k -j#{CREW_NPROC} check-clang"
-  #    system "make -k -j#{CREW_NPROC} check-lld"
-  #  end
+    #Dir.chdir("builddir") do
+      #system "ninja check-llvm"
+      #system "ninja check-clang"
+      #system "ninja check-lld"
+    #end
   #end
```

Comment:
It would be nice if we could get the tests working for such an important package.

Reply:
I can enable those. Those won't all complete.

Reply:
Here's the test output on x86_64:
```
[488/489] Running the LLVM regression tests
/usr/local/bin/ld.gold: /usr/lib64/libstdc++.so.6: version `CXXABI_1.3.9' not found (required by /usr/local/bin/ld.gold)
/usr/local/bin/ld.gold: /usr/lib64/libstdc++.so.6: version `GLIBCXX_3.4.21' not found (required by /usr/local/bin/ld.gold)
-- Testing: 38450 tests, 4 workers --
Testing:  0.. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80.. 90.. 

Testing Time: 880.90s
  Unsupported      :  1365
  Passed           : 36939
  Expectedly Failed:   146
```
```
********************
Testing:  0.. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80.. 90..
********************
Failed Tests (2):
  Clang :: Headers/arm-fp16-header.c
  Clang :: Headers/arm-neon-header.c


Testing Time: 853.83s
  Unsupported      :   107
  Passed           : 21268
  Expectedly Failed:    26
  Failed           :     2
```
```
-11.0.1-rc2.tar.gz.dir/llvm-project-llvmorg-11.0.1-rc2/builddir/bin/wasm-ld
-- Testing: 2432 tests, 4 workers --
Testing:  0.. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80.. 90..

Testing Time: 69.08s
  Unsupported      :   15
  Passed           : 2416
  Expectedly Failed:    1
Archiving...
llvm is built!
```
A lot of these unexpected errors are from arm tests on x86_64... so maybe not worth worrying about?

The GLIBC warning about ld.gold.... well... upgrading glibc is a PITA, right? (You've done it before... and the funny thing is that with LD_LIBRARY_PATH set to /usr/lib64 we could technically do it... since we're effectively almost doing a chroot...) 

I do have a coreutils_static package which works using musl if that is something that we want to do at some point...
```
require 'package'

class Coreutils_static < Package
  description 'GNU coreutils built statically with musl'
  homepage 'http://www.gnu.org/software/coreutils/coreutils.html'
  version '8.32'
  compatibility 'all'
  source_url 'https://mirrors.sarata.com/gnu/coreutils/coreutils-8.32.tar.xz'
  source_sha256 '4458d8de7849df44ccab15e16b1548b285224dbba5f08fac070c1c0e0bcc4cfa'


  def self.build
  # static build script from https://github.com/luciusmagn/coreutils-static/blob/master/build.sh
system 'cat <<"EOF"> build.sh
#!/bin/bash
#
# build static coreutils because we need exercises in minimalism
# MIT licensed: google it or see robxu9.mit-license.org.
#
# For Linux, also builds musl for truly static linking.

coreutils_version="8.32"
musl_version="1.2.1"

platform=$(uname -s)

if [ -d build ]; then
  echo "= removing previous build directory"
  rm -rf build
fi

#mkdir build # make build directory
#pushd build

# download tarballs
#echo "= downloading coreutils"
#curl -LO http://ftp.gnu.org/gnu/coreutils/coreutils-${coreutils_version}.tar.xz

#echo "= extracting coreutils"
#tar xJf coreutils-${coreutils_version}.tar.xz

if [ "$platform" = "Linux" ]; then
  echo "= downloading musl"
  curl -LO http://www.musl-libc.org/releases/musl-${musl_version}.tar.gz

  echo "= extracting musl"
  tar -xf musl-${musl_version}.tar.gz

  echo "= building musl"
  working_dir=$(pwd)

  install_dir=${working_dir}/musl-install

  pushd musl-${musl_version}
  env CFLAGS="$CFLAGS -Os -ffunction-sections -fdata-sections -fPIC" LDFLAGS=\'-Wl,--gc-sections\' ./configure --prefix=${install_dir}
  make install
  popd # musl-${musl-version}

  echo "= setting CC to musl-gcc"
  export CC=${working_dir}/musl-install/bin/musl-gcc
  export CFLAGS="-static"
else
  echo "= WARNING: your platform does not support static binaries."
  echo "= (This is mainly due to non-static libc availability.)"
fi

echo "= building coreutils"

#pushd coreutils-${coreutils_version}
env FORCE_UNSAFE_CONFIGURE=1 CFLAGS="$CFLAGS -Os -ffunction-sections -fdata-sections -fPIC" LDFLAGS=\'-Wl,--gc-sections\' ./configure
make
#popd # coreutils-${coreutils_version}

popd # build

if [ ! -d releases ]; then
  mkdir releases
fi

echo "= striptease"
strip -s -R .comment -R .gnu.version --strip-unneeded build/coreutils-${coreutils_version}/coreutils
#echo "= compressing"

shopt -s extglob
# upx currently breaks the binaries
#for file in build/coreutils-${coreutils_version}/src/!(*.*)
#do
#	upx --ultra-brute $file
#done
echo "= extracting coreutils binary"
cp src/!(*.*) releases
echo "= done"
EOF'
    system 'chmod +x ./build.sh'
    system './build.sh'
    system 'curl -Lf https://salsa.debian.org/debian/debianutils/-/raw/master/which?inline=false > releases/which && chmod +x releases/which'
    system "sed -i 's#/bin/sh#/usr/local/bin/sh#g' releases/which"

  end

  def self.install
    system "mkdir -p #{CREW_DEST_PREFIX}/bin"
    FileUtils.cd("#{CREW_DEST_PREFIX}/bin") do
      system "echo '#!/bin/bash' > arch"
      system "echo '#{CREW_PREFIX}/bin/uname -m' >> arch"
      system 'chmod +x arch'
    end
    system "cp releases/* #{CREW_DEST_PREFIX}/bin/"
  end
end
```